source/opt/def_use_manager.h

```diff
@@ -84,27 +87,50 @@ class DefUseManager {
   // information kept in this manager, but not the operands in the original
   // instructions.
   bool KillDef(uint32_t id);
+  // Turns the given instruction |inst| to a Nop, also erases both the use
+  // records of the result id of |inst| (if any) and the corresponding use
+  // records of: |inst| using |inst|'s operand ids.
+  void KillInst(ir::Instruction* inst);
   // Replaces all uses of |before| id with |after| id. Returns true if any
   // replacement happens. This method does not kill the definition of the
   // |before| id. If |after| is the same as |before|, does nothing and returns
   // false.
   bool ReplaceAllUsesWith(uint32_t before, uint32_t after);
 
  private:
-  // Clear the internal def-use record of a defined id if the given |def_id| is
-  // recorded by this manager. This method will erase both the uses of |def_id|
-  // and the |def_id|-generating instruction's use information kept in this
-  // manager, but not the operands in the original instructions.
+  // Clears the internal def-use records of a defined id if the given |def_id|
+  // is recorded by this manager. Does nothing if |def_id| has not been
+  // recorded yet. This method will erase both the uses of |def_id| and the
+  // |def_id|-generating instruction's use information kept in this manager,
+  // but not the operands in the original instructions.
   void ClearDef(uint32_t def_id);
 
-  using ResultIdToUsedIdsMap =
-      std::unordered_map<uint32_t, std::vector<uint32_t>>;
-
+  // Clears the internal def-use records of the given instruction |inst| if it
+  // has been analyzed by this manager. The use information of its operand ids
+  // will be updated: "the record: |inst| uses |operand id| will be removed".
+  // If the instruction is defining a result id, the uses of the result id will
+  // also be removed. Note that if |inst| has not been analyzed before, this
+  // function does nothing even though |inst| may define an existing result id.
+  void ClearInst(ir::Instruction* inst);
+
+  // Returns true if the operand's id should be recorded, otherwise returns
+  // false;
+  bool ShouldRecord(const ir::Operand& operand) const;
+
+  // Erases the record that: instruction |user| uses id |used_id|.
+  void EraseInstUseIdRecord(const ir::Instruction& user, uint32_t used_id);
+
+  // Resets the internal records
+  void Reset() {
+    analyzed_insts_.clear();
+    id_to_def_.clear();
+    id_to_uses_.clear();
+  }
+
+  std::unordered_set<ir::Instruction*>
+      analyzed_insts_;      // Analyzed instructions
```

Comment:
I don't think we need this map at all.  It just guards clearing of the def+uses of an instruction.  But the clear method should be resilient to the case where the instruction hasn't been analyzed before.
?!